# Library Management Analysis

A local library wants to better understand how its books are being borrowed. You are given a dataset containing information about books, members, and borrowing records. Your task is to analyze the dataset and produce useful insights that could help the library improve its services.

### Requirements

* Create a GitHub repository for the project and use Git throughout your development with meaningful commits.
* Include a `README.md` explaining the project, dataset, and how to run your code.
* Write clean, readable Python code and use functions whenever appropriate.
* Load the dataset using Pandas.
* Explore the dataset by displaying enough information to understand its structure before beginning the analysis.
* If the dataset contains any issues that prevent your analysis from running correctly, make only the minimum changes required to continue. You are not expected to perform a complete data cleaning process.
* Calculate the probability of randomly selecting:

  * A borrowed book from a specific genre.
  * A member belonging to a specific membership type.
  * A borrowing record that satisfies two conditions of your choice.
* Using Pandas, answer the following:

  * Find the category with the highest and lowest number of borrowings.
  * Determine which members borrow books most frequently.
  * Compare borrowing activity between different membership types.
  * Analyze how borrowing changes across publication years.
  * Identify the most popular authors.
  * Find the books that have never been borrowed.
  * Compare the average borrowing duration across genres.
  * Show the distribution of books across genres.
  * Analyze monthly borrowing trends (if borrowing dates are available).
  * Present three additional insights of your choice.
* Using NumPy, answer the following:

  * Convert one or more numerical columns into NumPy arrays.
  * Identify the highest and lowest values in a numerical column.
  * Use boolean indexing to filter a subset of the data.
  * Sort a NumPy array and display the result.
  * Perform at least one vectorized calculation to create a new numerical result or metric.
* Produce a final summary highlighting the most important insights from your analysis.

Your project should be well-structured and reproducible. Anyone running your code should be able to obtain the same results without modifying the source code.

In [ ]:
from google.colab import files

uploaded = files.upload()

Saving books.csv to books (2).csv
Saving borrowings.csv to borrowings.csv
Saving members.csv to members.csv


In [ ]:
#Load the dataset using Pandas.
import pandas as pd

books = pd.read_csv("books.csv")
borrowings = pd.read_csv("borrowings.csv")
members = pd.read_csv("members.csv")

In [ ]:
#Explore the dataset by displaying enough information to understand its structure before beginning the analysis.
print("**************** FOR BOOKS ****************")
print(books)
print(books.info())

print("**************** FOR BORROWINGS ****************")
print(borrowings)
print(borrowings.info())

print("**************** FOR MEMBERS ****************")
print(members)
print(members.info())

**************** FOR BOOKS ****************
    BookID     Title             Author            Genre  PublicationYear  \
0     B001    Book 1      George Orwell  Science Fiction             1986   
1     B002    Book 2  Brandon Sanderson          Mystery             1991   
2     B003    Book 3        Jane Austen      Programming             1987   
3     B004    Book 4  Brandon Sanderson         Business             2023   
4     B005    Book 5           Mark Lee         Business             2011   
..     ...       ...                ...              ...              ...   
395   B396  Book 396  Brandon Sanderson         Business             2002   
396   B397  Book 397       J.K. Rowling        Self Help             2022   
397   B398  Book 398      George Orwell         Business             1999   
398   B399  Book 399       J.K. Rowling      Programming             2006   
399   B400  Book 400      George Orwell          Mystery             1985   

         Publisher  Pages Langu

In [ ]:
# the probability of randomly selecting A borrowed book from a specific genre
total_books = len(books)
book_genres = books['Genre'].unique()
print(f"the generies is {book_genres}")
for genre in book_genres :
  total_specific = len(books[books["Genre"] == genre])
  probability = (total_specific / total_books) * 100
  print(f"the probability of randomly selecting A borrowed book from a {genre} genre is {probability}%")

print("\n\n")

# the probability of randomly selecting A member belonging to a specific membership type.
total_members = len(members)
members_type = members['MembershipType'].unique()
print(f"the memberships is {members_type}")
for member_type in members_type:
  total_specific = len(members[members['MembershipType'] == member_type])
  probability =  (total_specific / total_members) * 100
  print(f"the probability of randomly selecting A member belonging to a {member_type} membership type is {probability:0.1f}%")

print("\n\n")

#the probability of randomly selecting A borrowing record that satisfies two conditions of your choice.
#I set two conditions -> borrowed in after 2024 and returned before 2026
total_borrowing = len(borrowings)
borrowings['BorrowDate'] = pd.to_datetime(borrowings['BorrowDate'])
borrowings['ReturnDate'] = pd.to_datetime(borrowings['ReturnDate'])
total_specific = len(borrowings[(borrowings['BorrowDate'].dt.year > 2024) & (borrowings['ReturnDate'].dt.year  < 2026)])
probability = (total_specific / total_borrowing) * 100
print(f"the probability of randomly selecting A borrowing record that borrowed in after 2024 and returned before 2026 is {probability}")

the generies is ['Science Fiction' 'Mystery' 'Programming' 'Business' 'Fantasy'
 'Data Science' 'History' 'Romance' 'Biography' 'Self Help']
the probability of randomly selecting A borrowed book from a Science Fiction genre is 8.75%
the probability of randomly selecting A borrowed book from a Mystery genre is 10.75%
the probability of randomly selecting A borrowed book from a Programming genre is 10.0%
the probability of randomly selecting A borrowed book from a Business genre is 13.0%
the probability of randomly selecting A borrowed book from a Fantasy genre is 8.0%
the probability of randomly selecting A borrowed book from a Data Science genre is 10.75%
the probability of randomly selecting A borrowed book from a History genre is 9.75%
the probability of randomly selecting A borrowed book from a Romance genre is 8.5%
the probability of randomly selecting A borrowed book from a Biography genre is 10.0%
the probability of randomly selecting A borrowed book from a Self Help genre is 10.

In [ ]:
borrowings_books_table = pd.merge(books, borrowings, how = 'inner', on = 'BookID')
borrowings_members_table = pd.merge(borrowings, members, how='inner', on='MemberID')
borrowings_members_books_table = pd.merge(books, borrowings_members_table, how='inner', on='BookID')

,BookID,Title,Author,Genre,PublicationYear,Publisher,Pages,Language,CopiesAvailable,BorrowID,MemberID,BorrowDate,ReturnDate,Name,MembershipType,Age,Gender,JoinDate
0,B001,Book 1,George Orwell,Science Fiction,1986,Pearson,401,English,3,BR0095,M025,2024-04-03,2024-04-28,Member 25,Student,56,Male,2024-07-03
1,B001,Book 1,George Orwell,Science Fiction,1986,Pearson,401,English,3,BR0778,M102,2025-04-08,2025-04-25,Member 102,Student,31,Female,2023-07-10
2,B002,Book 2,Brandon Sanderson,Mystery,1991,Pearson,678,English,1,BR0080,M149,2025-05-14,2025-05-27,Member 149,Faculty,38,Male,2024-03-26
3,B002,Book 2,Brandon Sanderson,Mystery,1991,Pearson,678,English,1,BR0144,M131,2024-04-28,2024-05-08,Member 131,Faculty,56,Female,2024-02-28
4,B002,Book 2,Brandon Sanderson,Mystery,1991,Pearson,678,English,1,BR0422,M057,2026-02-16,2026-03-12,Member 57,Faculty,33,Male,2024-03-23
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,B400,Book 400,George Orwell,Mystery,1985,Random House,281,English,6,BR1240,M127,2026-02-22,2026-03-05,Member 127,Public,29,Female,2024-07-08
1996,B400,Book 400,George Orwell,Mystery,1985,Random House,281,English,6,BR1293,M004,2025-01-11,2025-01-19,Member 4,Student,65,Male,2025-03-30
1997,B400,Book 400,George Orwell,Mystery,1985,Random House,281,English,6,BR1422,M056,2024-07-30,2024-08-21,Member 56,Faculty,65,Male,2024-09-07
1998,B400,Book 400,George Orwell,Mystery,1985,Random House,281,English,6,BR1431,M091,2024-08-28,2024-09-15,Member 91,Premium,35,Female,2022-06-11


In [ ]:
#Using Pandas, answer the following:
#Find the category with the highest and lowest number of borrowings.
category_borrowings = borrowings_books_table.groupby('Genre')['BorrowID'].count().sort_values(ascending=False)
print(category_borrowings)
print(f"So the category with the highest number of borrowings is {category_borrowings.idxmax()} with {category_borrowings.max()} borrowed books and the lowest is {category_borrowings.idxmin()} with {category_borrowings.min()} borrowed books.")
print("***********************************")

#Determine which members borrow books most frequently.
print(borrowings_members_table.groupby('Name')['BorrowID'].count().sort_values(ascending=False))

print("***********************************")

#Compare borrowing activity between different membership types.
print(borrowings_members_table.groupby('MembershipType')['BorrowID'].count().sort_values(ascending=False))

print("***********************************")

Genre
Mystery            512
Data Science       328
Fantasy            316
Programming        295
Business           121
Self Help           97
Biography           84
History             83
Science Fiction     83
Romance             81
Name: BorrowID, dtype: int64
So the category with the highest number of borrowings is Mystery with 512 borrowed books and the lowest is Romance with 81 borrowed books.
***********************************
Name
Member 78     31
Member 147    24
Member 122    23
Member 126    23
Member 112    20
              ..
Member 136     6
Member 76      6
Member 108     5
Member 29      5
Member 121     4
Name: BorrowID, Length: 150, dtype: int64
***********************************
MembershipType
Student    1197
Public      415
Faculty     289
Premium      99
Name: BorrowID, dtype: int64
***********************************


In [ ]:
#Analyze how borrowing changes across publication years.
print(borrowings_books_table.groupby('PublicationYear')['BorrowID'].count())

print("***********************************")

#Identify the most popular authors.
author_borrowing = borrowings_books_table.groupby('Author')['BorrowID'].count().sort_values(ascending=False)
print(author_borrowing)
print(f'So the most popular authors is {author_borrowing.idxmax()} with {author_borrowing.max()} borrowed books')


PublicationYear
1985     40
1986     19
1987     90
1988     20
1989    108
1990     86
1991     48
1992     34
1993     24
1994     62
1995     45
1996     35
1997     48
1998     62
1999     15
2000     59
2001     58
2002     70
2003    102
2004     41
2005     58
2006     44
2007     81
2008     46
2009     12
2010     23
2011     57
2012     30
2013     44
2014     15
2015     40
2016     50
2017     40
2018     46
2019     96
2020     50
2021     38
2022     45
2023     44
2024     24
2025     51
Name: BorrowID, dtype: int64
***********************************
Author
Alice Brown          191
David Clark          178
Sarah Mitchell       170
Brandon Sanderson    169
J.K. Rowling         157
Yuval Noah Harari    133
Robert Martin        126
John Smith           121
Maria Green          119
Emily Stone          118
Mark Lee             116
George Orwell        113
Isaac Asimov         103
Jane Austen          103
Agatha Christie       83
Name: BorrowID, dtype: int64
So the most popu

In [ ]:
#Find the books that have never been borrowed.
merge = pd.merge(books, borrowings, how = 'left', on = 'BookID')
book_nborrowwed = merge[merge['BorrowID'].isna()]
print(book_nborrowwed['Title'])

print("***********************************")

#Compare the average borrowing duration across genres.
borrowings_books_table['Borrowing Duration'] = borrowings_books_table['ReturnDate'] - borrowings_books_table['BorrowDate']
print(borrowings_books_table.groupby('Genre')['Borrowing Duration'].mean().sort_values(ascending=False))

70       Book 12
71       Book 13
74       Book 15
139      Book 31
156      Book 39
193      Book 44
451      Book 92
608     Book 125
626     Book 129
856     Book 175
906     Book 181
1075    Book 213
1177    Book 242
1303    Book 268
1437    Book 294
1444    Book 298
1512    Book 310
1535    Book 317
1569    Book 322
1654    Book 337
1684    Book 344
1703    Book 348
1916    Book 378
Name: Title, dtype: object
***********************************
Genre
Biography         18 days 03:08:34.285714285
Science Fiction   17 days 19:05:03.614457831
Self Help         17 days 12:51:57.525773196
Data Science      17 days 11:20:29.268292683
Programming       17 days 04:57:45.762711864
Business          17 days 04:33:43.140495867
Mystery              16 days 22:24:22.500000
Romance           16 days 20:44:26.666666666
History           16 days 11:51:19.518072289
Fantasy           15 days 19:31:08.354430379
Name: Borrowing Duration, dtype: timedelta64[ns]


In [ ]:
#Show the distribution of books across genres.
print(books.groupby('Genre')['BookID'].count().sort_values(ascending = False))

print("***********************************")

#Analyze monthly borrowing trends (if borrowing dates are available).
borrowings_books_table['Borrowing Month'] = borrowings_books_table['BorrowDate'].dt.month
print(borrowings_books_table.groupby('Borrowing Month')['BorrowID'].count())

print("***********************************")

#Present three additional insights of your choice
#Which age group borrows the most books
print(borrowings_members_table.groupby('Age')['BorrowID'].count().sort_values(ascending = False).head())

print("***********************************")

#Average number of pages by genre
print(books.groupby('Genre')['Pages'].mean().astype(int).sort_values(ascending = False))

print("***********************************")

#Which genre has the most total copies available
print(books.groupby('Genre')['CopiesAvailable'].sum().sort_values(ascending = False))

Genre
Business           52
Data Science       43
Mystery            43
Self Help          42
Biography          40
Programming        40
History            39
Science Fiction    35
Romance            34
Fantasy            32
Name: BookID, dtype: int64
***********************************
Borrowing Month
1     208
2     193
3     218
4     200
5     204
6     161
7     153
8     142
9     115
10    126
11    147
12    133
Name: BorrowID, dtype: int64
***********************************
Age
53    96
38    80
24    79
59    76
31    71
Name: BorrowID, dtype: int64
***********************************
Genre
Mystery            461
History            447
Science Fiction    437
Business           427
Self Help          425
Data Science       417
Fantasy            412
Biography          396
Romance            395
Programming        394
Name: Pages, dtype: int64
***********************************
Genre
Business           205
Self Help          200
Biography          182
Data Science       173


In [ ]:
import numpy as np

#Convert one or more numerical columns into NumPy arrays.
books_pages = np.array(books['Pages'])
print(books_pages)

print("******************************************************************************************")

books_copies = np.array(books['CopiesAvailable'])
print(books_copies)

print("******************************************************************************************")

#Identify the highest and lowest values in a numerical column.
print(f"The highest number of pages in one book is {np.max(books_pages)} and the lowest is {np.min(books_pages)}.")

print("******************************************************************************************")

#Use boolean indexing to filter a subset of the data.
big_books = books_pages[books_pages > 500]
print(len(big_books))

print("******************************************************************************************")

#Sort a NumPy array and display the result.
sorted_books_pages = np.sort(books_pages)
print(sorted_books_pages)

print("******************************************************************************************")

#Perform at least one vectorized calculation to create a new numerical result or metric.
number_of_papers = books_pages * books_copies
print(number_of_papers)

[401 678 215 694 579 552 224 390 507 490 353 404 483 295 508 452 394 631
 262 558 625 283 599 237 420 389 425 285 451 365 617 606 553 439 582 141
 192 192 672 368 219 175 374 552 573 215 338 122 689 179 608 640 189 160
 443 525 194 195 477 281 661 226 686 471 377 123 285 234 498 435 375 278
 145 374 601 353 405 641 238 231 643 165 321 441 538 316 428 335 449 604
 463 323 194 513 125 650 587 444 576 373 364 398 274 544 183 140 519 619
 518 250 147 306 585 378 478 373 364 596 237 439 368 699 499 621 276 671
 450 369 149 337 402 367 607 346 497 683 377 668 657 489 400 249 600 378
 195 207 351 559 221 203 280 418 584 322 272 289 247 632 160 146 395 572
 623 448 627 585 533 543 630 682 616 282 215 277 399 587 640 468 425 615
 586 512 604 249 222 267 391 666 190 212 426 487 541 511 592 414 426 323
 510 525 628 344 219 493 545 663 613 197 494 669 510 694 184 263 219 584
 278 537 575 496 402 337 698 341 403 232 451 551 490 346 151 230 449 681
 584 237 353 217 276 562 181 482 439 278 496 559 69

## Final Summary: Key Insights from the Library Management Analysis

**Dataset overview:** 400 books, 2,000 borrowing records, and 150 members across 4 membership types and 10 genres.

### Borrowing patterns
- **Mystery is the most-borrowed genre by far** (512 borrowings), followed by Data Science (328) and Fantasy (316). **Romance is the least borrowed** (81), despite not being the smallest category in the catalog — an early sign that catalog size and borrowing demand don't move together.
- **Students drive the vast majority of activity**: 1,197 of 2,000 borrowings (~60%), consistent with Students also being 60% of the membership base. Public members (20% of members) account for 415 borrowings, Faculty (14.7%) account for 289, and Premium members (5.3%, the smallest group) account for just 99 — roughly proportional to their share of membership, suggesting no single membership type borrows disproportionately more *per member* than expected from its size alone.
- **Borrowing activity is highly concentrated among a few power users**: Member 78 alone borrowed 31 books, compared to a long tail of members who borrowed only 4–6 books. This kind of concentration is worth flagging for engagement programs.
- **Alice Brown is the most borrowed author** (191 borrowings), narrowly ahead of David Clark (178) and Sarah Mitchell (170); the spread across the top 5 authors is fairly tight, indicating broad author appeal rather than reliance on one star author.
- **Borrowing has a clear seasonal dip**: monthly borrowings peak in March (218) and decline steadily through the year to a low in September (115) — roughly a 47% drop from peak to trough. This points to a strong case for targeted promotions or events in the slower autumn months.

### Catalog and inventory insights
- **23 of 400 books (~5.75%) have never been borrowed**, including titles like *Book 12*, *Book 31*, and *Book 213*. These are candidates for promotion, repositioning, or eventual de-acquisition.
- **Genre popularity and catalog size are mismatched.** Business has the largest catalog (52 books) but ranks only 5th in borrowings (121), while Mystery has a mid-sized catalog (43 books) yet generates the most borrowings (512) — over 4x more demand per book than Business. This is the clearest actionable insight in the dataset: **the library may be over-stocked in Business/Self Help and under-stocked in Mystery/Data Science relative to demand.**
- **Borrowing duration is fairly uniform across genres** (15.8–18.1 days), with Biography borrowed longest and Fantasy returned fastest — a relatively narrow ~2-day spread suggests loan length is driven more by library policy/due dates than by genre-specific reading habits.
- Average book length ranges from 394 pages (Romance) to 461 pages (Mystery) — genre doesn't strongly predict book length, and page count alone doesn't explain why Mystery dominates borrowings.

### Member behavior
- **Borrowing is fairly evenly spread across age groups**, with the busiest single ages (53, 38, 24, 59, 31) each contributing similar counts — there's no strong evidence that one age bracket dominates usage.
- The probability calculations show genre selection probabilities ranging narrowly from 8% to 13%, meaning **no single genre dominates the catalog composition** — the borrowing skew toward Mystery is a demand effect, not a supply effect.

### Recommendations for the library
1. **Rebalance acquisitions**: shift future purchasing toward Mystery and Data Science, and slow down acquisitions in Business and Self Help, where supply already outpaces demand.
2. **Run a "rediscover" campaign** for the 23 never-borrowed titles (featured displays, staff picks, recommendation engine boosts) before considering removing them from the catalog.
3. **Target retention efforts at Student members**, since they generate the majority of borrowings — small improvements in their experience would have an outsized impact on overall library usage.
4. **Plan for the September–October slowdown** with seasonal promotions, reading challenges, or events to flatten the year-over-year dip in borrowing activity.
5. **Investigate and recognize top borrowers** (e.g., Member 78) — understanding what drives their high engagement could inform strategies to lift the long tail of lower-activity members.

